# <center>Đồ Án Cuối Kì</center>

## Mục lục
* [Danh sách thành viên](#c1)
* [Cách chạy Project](#c2)
* [Nội dung](#c3)
    * [Thu thập dữ liệu](#c31)
        * [Chủ đề dữ liệu của bạn là gì? Nguồn lấy từ đâu?](#c311)
        * [Bạn có được quyền sử dụng dữ liệu từ tác giả hay không?](#c312)
        * [Tác giả thu thập dữ liệu như thế nào?](#c313)
    * [Khám phá dữ liệu](#c32)
        * [Dữ liệu có bao nhiêu dòng và bao nhiêu cột?](#c321)
        * [Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?](#c322)
        * [Dữ liệu có các dòng bị lặp không?](#c323)
        * [Mỗi cột có ý nghĩa gì?](#c324)
        * [Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?](#c325)
        * [Với mỗi cột có kiểu dữ liệu dạng số (numeric), các giá trị được phân bố như thế nào?](#c325)
        * [Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?](#c326)
    * [Đưa ra câu hỏi có ý nghĩa cần trả lời](#c33)
* [Nguồn tham khảo](#c4)

## Danh sách thành viên <a class="anchor" id="c1"></a>

STT&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;Thông tin sinh viên&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&ensp;Github username
<br>
1&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;Phạm Ngọc Thiên Ân - 19127094&emsp;&emsp;&emsp;&ensp;PNTAN19
<br>
2&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;Phan Vĩ Giai - 19127639&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;owfdataboy

## CÁCH CHẠY PROJECT <a class="anchor" id="c2"></a>
**Chọn Kernel -> Restart & Run All**

## Nội dung <a class="anchor" id="c3"></a>

### Import các thư viện cần thiết

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

### Đọc dữ liệu vào DataFrame

In [2]:
df = pd.read_csv("StudentsPerformance.csv")

### Thu thập dữ liệu <a class="anchor" id="c31"></a>

### Chủ đề dữ liệu của bạn là gì? Nguồn lấy từ đâu? <a class="anchor" id="c311"></a>

Trong đồ án này, nhóm em sẽ thực hiện phân tích dữ liệu về "Những tác nhân tác động lên điểm số của học sinh" ở nước Mỹ.

Dữ liệu tên là "Students Performance in Exams" của tác giả Jakki Seshapanpu, ông sinh sống tại Ấn độ.\
Dữ liệu được tác giả public trên Kaggle.com, được download [tại đây](https://www.kaggle.com/spscientist/students-performance-in-exams).

### Bạn có được quyền sử dụng dữ liệu từ tác giả hay không? <a class="anchor" id="c312"></a>

### Tác giả thu thập dữ liệu như thế nào? <a class="anchor" id="c313"></a>

### Khám phá dữ liệu <a class="anchor" id="c32"></a>

### Dữ liệu có bao nhiêu dòng và bao nhiêu cột? <a class="anchor" id="c321"></a>

In [3]:
df.shape

(1000, 8)

Dữ liệu gồm 1000 dòng (tương đương với 1000 học sinh) và 8 cột (8 trường dữ liệu)

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không? <a class="anchor" id="c322"></a>

Theo file mô tả dữ liệu và theo quan sát sơ bộ về dữ liệu thì một dòng cho biết thông tin của các thuộc tính và số điểm của học sinh. Và các dòng có ý nghĩa tương đương nhau.

### Dữ liệu có các dòng bị lặp không? <a class="anchor" id="c323"></a>

In [4]:
have_duplicated_rows = df.duplicated().eq(True).all()
have_duplicated_rows

False

Dữ liệu không bị trùng lặp

### Mỗi cột có ý nghĩa gì? <a class="anchor" id="c324"></a>

Sau đây là phần mô tả dữ liệu cho "Students Performance in Exams":
- **gender**: Giới tính của học sinh
- **race/ethnicity**: Chủng tộc/dân tộc của học sinh
- **parental level of education**: Trình độ học vấn của cha mẹ
- **lunch**: Mức độ bữa ăn của học sinh
- **test preparation course**: Sự chuẩn bị cho bài kiểm tra
- **math score**: Điểm môn toán của học sinh
- **reading score**: Điểm môn đọc của học sinh
- **writing score**: Điểm môn viết của học sinh

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không? <a class="anchor" id="c325"></a>

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


Theo như quan sát dữ liệu trên thì có thể thấy:
- Có 5 cột mang dữ liệu `object` gồm cột "gender", "race/ethnicity", "parental level of education", "lunch" và "test preparation course".
- Và có 3 cột mang kiểu dữ liệu `int64` là các cột điểm của chúng ta gồm "math score", "reading score" và "writing score"

Ở đây, có kiểu dữ liệu `object` khá bất thường ? Trong Pandas, kiểu dữ liệu object thường ám chỉ chuỗi, nhưng thật ra kiểu dữ liệu object có thể chứa một đối tượng bất kỳ trong Python (vì thật ra ở bên dưới kiểu dữ liệu object chứa địa chỉ). Nếu cột có dữ liệu là `object` thì kiểu dữ liệu của cột có thể khác nhau. Để có thể chắc chắn 5 cột dữ liệu `object` của chúng ta thuộc kiểu dữ liệu mong muốn là `str` thì chúng ta phải thực hiện một thao tác kiểm tra từng phần tử của cột đó, để có thể biết chắc rằng thực sự cột đó có thể chứa những kiểu dữ liệu nào khác hay không ?

In [6]:
def open_object_dtype(s):
    dtypes = set()
    
    def get_type(cell):
        if type(cell) is str:
            dtypes.add(str)
        elif type(cell) is float:
            dtypes.add(float)
    s.apply(get_type)
    
    return dtypes

In [7]:
assert open_object_dtype(df['gender']) == {str}
assert open_object_dtype(df['race/ethnicity']) == {str}
assert open_object_dtype(df['parental level of education']) == {str}
assert open_object_dtype(df['lunch']) == {str}
assert open_object_dtype(df['test preparation course']) == {str}

Kết quả cuối cùng cho thấy các cột có kiểu dữ liệu đúng như mong đợi của chúng ta. Nên sẽ không cần xử lý về kiểu dữ liệu cho các cột nữa.

### Với mỗi cột có kiểu dữ liệu dạng số (numeric), các giá trị được phân bố như thế nào?
<a class="anchor" id="c326"></a>

### Dữ liệu thống kê cho thuộc tính dạng số

In [11]:
df.describe().T

count    mean        std   min    25%   50%   75%    max
math score     1000.0  66.089  15.163080   0.0  57.00  66.0  77.0  100.0
reading score  1000.0  69.169  14.600192  17.0  59.00  70.0  79.0  100.0
writing score  1000.0  68.054  15.195657  10.0  57.75  69.0  79.0  100.0

#### Một chút nhận xét về dữ liệu thống kê:

Dữ liệu số bao gồm 3 cột `math score`, `reading score`, `writing score` đều có 1000 dòng.

Đối với thuộc tính `math score`:
- Trung bình rơi vào khoảng `66` điểm
- Độ lệch chuẩn là `15.163`
- Điểm thấp nhất của `môn toán ` là 0 điểm
- Điểm cao nhất của `môn toán ` là 100 điểm
- Thông số thống kê `Interquartile`: 
    - Có 25% học sinh thuộc loại trung bình yếu (< 57 điểm)
    - Có 50% học sinh trên trung bình (trên 66 điểm)
    - Và có 75% học sinh thuộc loại giỏi (trên 77 điểm)
    
Tương tự đối với thuộc tính `reading score`:
- Trung bình rơi vào khoảng `69` điểm
- Độ lệch chuẩn là `14.6`
- Điểm thấp nhất của `môn đọc ` là 17 điểm
- Điểm cao nhất của `môn đọc ` là 100 điểm
- Thông số thống kê `Interquartile`: 
    - Có 25% học sinh thuộc loại trung bình yếu (< 59 điểm)
    - Có 50% học sinh trên trung bình (trên 70 điểm)
    - Và có 75% học sinh thuộc loại giỏi (trên 79 điểm)
    
Tương tự đối với thuộc tính `writing score`:
- Trung bình rơi vào khoảng `68` điểm
- Độ lệch chuẩn là `15.195`
- Điểm thấp nhất của `môn viết ` là 10 điểm
- Điểm cao nhất của `môn đọc ` là 100 điểm
- Thông số thống kê `Interquartile`: 
    - Có 25% học sinh thuộc loại trung bình yếu (< 57.75 điểm)
    - Có 50% học sinh trên trung bình (trên 69 điểm)
    - Và có 75% học sinh thuộc loại giỏi (trên 79 điểm)
    
So sánh dữ liệu điểm của ba thuộc tính `math score`, `reading score`, `writing score`:
- Cả ba môn đều có học sinh đạt số điểm tối đa là 100 (Quá giỏi :)) )
- Cả ba môn đều có số học sinh khá/giỏi (Đạt số điểm > 75 điểm) khá tương tự nhau => Lớp này có thể có các học sinh học đồng đều tất cả ba môn không ? Maybe...
- Dựa vào `giá trị trung bình` và `độ lệch chuẩn` ta có thể thấy sự dàn trải dữ liệu `(range)` của ba môn đều khá đều như nhau 
- `Điểm toán` là môn có học sinh 0 điểm (Trong lớp có học sinh yếu toán :>)
- Chúng ta cần trực quan dữ liê

#### Dữ liệu điểm của từng môn được phân bố như thế nào ?

### Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?
<a class="anchor" id="c327"></a>

### Đưa ra câu hỏi có ý nghĩa cần trả lời <a class="anchor" id="c33"></a>

## Nguồn tham khảo <a class="anchor" id="c4"></a>